src/herder/PendingEnvelopes.h

```diff
@@ -65,14 +70,39 @@ class PendingEnvelopes
 
     // discards all SCP envelopes thats use QSet with given hash,
     // as it is not sane QSet
-    void discardSCPEnvelopesWithQSet(Hash hash);
+    void discardSCPEnvelopesWithQSet(Hash const& hash);
 
     void updateMetrics();
 
+    void envelopeReady(SCPEnvelope const& envelope);
+    void discardSCPEnvelope(SCPEnvelope const& envelope);
+    bool isFullyFetched(SCPEnvelope const& envelope);
+    void startFetch(SCPEnvelope const& envelope);
+    void stopFetch(SCPEnvelope const& envelope);
+    void touchFetchCache(SCPEnvelope const& envelope);
+    bool isDiscarded(SCPEnvelope const& envelope) const;
+
+    SCPQuorumSetPtr putQSet(Hash const& qSetHash, SCPQuorumSet const& qSet);
+    // tries to find a qset in memory, setting touch also touches the LRU,
+    // extending the lifetime of the result
+    SCPQuorumSetPtr getKnownQSet(Hash const& hash, bool touch);
+
+    // NB: caller should use the return value to avoid duplicate tx sets
+    TxSetFramePtr putTxSet(Hash const& hash, uint64 slot, TxSetFramePtr txset);
+    // tries to find a txset in memory, setting touch also touches the LRU,
+    // extending the lifetime of the result
+    TxSetFramePtr getKnownTxSet(Hash const& hash, uint64 slot, bool touch);
+
+    void cleanKnownData();
```

Comment:
I would strongly discourage use of `std::shared_ptr<TxSetFrame>(new TxSetFrame(txset))` because it can leak memory. If we _do_ want to avoid the control block issue, then we should allocate the pointer in a RAII fashion before transferring control to the `shared_ptr`.

Reply:
I left this with the control block for now as the shared ptrs are of `TxSetFrame` and `SCPQuorumSet`, both of which are fairly small (they contain vectors that are destroyed right away so we're left with a few bytes)